src/video-player/components/runtime.tsx

```diff
@@ -0,0 +1,155 @@
+// sample captions
+// "https://models-resources.s3.amazonaws.com/question-interactives/test-captions.vtt";
+
+let saveStateInterval : any;
+
+export const Runtime: React.FC<IProps> = ({ authoredState, interactiveState, setInteractiveState, setNavigation, report }) => {
+  let viewedProgress = interactiveState?.percentageViewed || 0;
+  let viewedTimestamp = interactiveState?.lastViewedTimestamp || 0;
+  const playerRef = useRef(null);
+  useEffect(() => {
+    loadPlayer();
+    Shutterbug.enable("." + css.runtime);
+    Shutterbug.on("saycheese", playerRef.current);
+    return () => {
+      Shutterbug.off("saycheese", playerRef.current);
+      Shutterbug.disable();
+    };
+  }, []);
+  const getViewTime = () => {
+    if (playerRef.current) {
+      const video: HTMLVideoElement = playerRef.current! as HTMLVideoElement;
+      return video.currentTime;
+    }
+    else return 0;
+  };
+  const getViewPercentage = () => {
+    if (playerRef.current) {
+      const video: HTMLVideoElement = playerRef.current! as HTMLVideoElement;
+      return video.currentTime / video.duration;
+    }
+    else return 0;
+  };
+
+  const { submitButton, lockedInfo } = useRequiredQuestion({ authoredState, interactiveState, setInteractiveState, setNavigation, isAnswered: viewedProgress > 0.96 });
+
+  const loadPlayer = () => {
+    const player: videojs.Player = videojs(playerRef.current,
+      {
+        controls: true,
+        fluid: authoredState.fixedAspectRatio || authoredState.fixedHeight ? false : true
+      }, () => {
+        const url = authoredState.videoUrl ? authoredState.videoUrl : "";
+        player.src(url);
+
+        if (authoredState.captionUrl) {
+          player.addRemoteTextTrack({
+            kind: 'captions',
+            language: 'en',
+            label: 'English',
+            src: authoredState.captionUrl,
+            'default': true
+          }, false);
+        }
+      });
+
+    if (authoredState.fixedAspectRatio) {
+      player.aspectRatio(getAspectRatio(authoredState.fixedAspectRatio));
+    }
+    if (authoredState.fixedHeight) {
+      player.height(authoredState.fixedHeight);
+    }
+    player.ready(() => {
+      if (viewedTimestamp) {
+        // console.log(player.duration());
+        // console.log(viewedProgress);
+        player.currentTime(viewedTimestamp);
+      }
+    });
+
+    return () => {
+      player.dispose();
+    };
+  };
+
+  const getAspectRatio = (aspectRatio: string) => {
+    if (aspectRatio.indexOf(':') > -1) {
+      // user supplied aspect ratio as a:b, so use it
+      return aspectRatio;
+    } else {
+      // numeric - so make it into a:b format
+      const roundedAspect = Math.round(parseFloat(aspectRatio) * 100);
+      return `${roundedAspect}:100`;
+    }
+  };
+
+  const handlePlaying = (e: any) => {
+    // store current playback progress each second
+    saveStateInterval = setInterval(updateState, 1000);
```

Comment:
Using module level variable for `saveStateInterval` will work in this case, as we'll have only one Runtime pretty much always. But generally, it might not be the best/safest assumption. In such case it's nice to use refs:
```
// At the beginning of the functional component function:
const saveStateIntervalRef = useRef<number>();
// Here:
saveStateInterval.current = setInterval(updateState, 1000);
```
Or state, but ref is better, as a change of this value shouldn't trigger re-render.